stan/math/prim/mat/fun/mdivide_left_tri.hpp

```diff
@@ -52,6 +58,77 @@ inline Eigen::Matrix<T, R1, C1> mdivide_left_tri(
   return b;
 }
 
+/**
+ * Returns the solution of the system Ax=b when A is triangular
+ * and A and b are matrices of doubles.
+ * @param A Triangular matrix.  Specify upper or lower with TriView
+ * being Eigen::Upper or Eigen::Lower.
+ * @param b Right hand side matrix or vector.
+ * @return x = A^-1 b, solution of the linear system.
+ * @throws std::domain_error if A is not square or the rows of b don't
+ * match the size of A.
+ */
+template <int TriView, int R1, int C1, int R2, int C2>
+inline Eigen::Matrix<double, R1, C2> mdivide_left_tri(
+    const Eigen::Matrix<double, R1, C1> &A,
+    const Eigen::Matrix<double, R2, C2> &b) {
+  check_square("mdivide_left_tri", "A", A);
+  check_multiplicable("mdivide_left_tri", "A", A, "b", b);
+#ifdef STAN_OPENCL
+  if (A.rows()
+      >= opencl_context.tuning_opts().lower_tri_inverse_size_worth_transfer) {
+    matrix_cl A_cl(A);
+    matrix_cl b_cl(b);
+    matrix_cl A_inv_cl(A.rows(), A.cols());
+    if (TriView == Eigen::Lower) {
+      A_inv_cl = lower_triangular_inverse(A_cl);
+    } else {
+      A_inv_cl = transpose(lower_triangular_inverse(transpose(A_cl)));
+    }
+    matrix_cl C_cl = A_inv_cl * b_cl;
+    return from_matrix_cl(C_cl);
+  } else {
+#endif
+    return A.template triangularView<TriView>().solve(b);
+#ifdef STAN_OPENCL
+  }
+#endif
+}
+
+/**
+ * Returns the solution of the system Ax=b when A is triangular, b=I and
+ * both are matrices of doubles.
+ * @param A Triangular matrix.  Specify upper or lower with TriView
+ * being Eigen::Upper or Eigen::Lower.
+ * @return x = A^-1 .
+ * @throws std::domain_error if A is not square
+ */
+template <int TriView, int R1, int C1>
+inline Eigen::Matrix<double, R1, C1> mdivide_left_tri(
+    const Eigen::Matrix<double, R1, C1> &A) {
+  check_square("mdivide_left_tri", "A", A);
+  const int n = A.rows();
+#ifdef STAN_OPENCL
```

Comment:
Not for this PR, but we need to find a solution for to get rid of these ifdefs. Maybe we could just set STAN_OPENCL to 0 by default, then if we always do the check for transfer we could do something like 

```cpp
...
  if (STAN_OPENCL && A.rows()
      >= opencl_context.tuning_opts().lower_tri_inverse_size_worth_transfer) {
...
```
We'd probably have to make a dummy opencl_context or something, but the compiler would just get rid of the if statement above if STAN_OPENCL is 0

Reply:
Yeah a dummy context seems the best idea to me. But would hold off that cleanup for another PR, maybe after we deal with this first batch of things (mdivide, multiply, GLMs).